### PR TITLE
Added the Blazor Viewer

### PR DESCRIPTION
**Pull request by jaimeatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WEB-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-WEB [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WEB-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
The Viewer developed in Blazor has been included in COMET-WEB-Community-Edition

### CODE REVIEW

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Canvas on mouse down event
+        /// </summary>
+        /// <param name="e">the mouse args of the event</param>
+        public void OnMouseDown(MouseEventArgs e)
+        {
+
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:96]: I propose you put a `//TODO:` statement here to make it explicit what you intend to code here.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        /// <summary>
+        /// Canvas on mouse up event
+        /// </summary>
+        /// <param name="e">the mouse args of the event</param>
+        public void OnMouseUp(MouseEventArgs e)
+        {
+
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:105]: I propose you put a `//TODO:` statement here to make it explicit what you intend to code here.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        /// <summary>
+        /// Canvas on click event
+        /// </summary>
+        /// <param name="e">the mouse args of the event</param>
+        public void OnClick(MouseEventArgs e)
+        {
+
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:114]: I propose you put a `//TODO:` statement here to make it explicit what you intend to code here.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            Scene.SetInfoPanelVisibility(false);
+        }
+
+        /// <summary>
+        /// Canvas on key down event
+        /// </summary>
+        /// <param name="e">the keyboard args of the event</param>
+        public void OnKeyDown(KeyboardEventArgs e)
+        {
+
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:141]: I propose you put a `//TODO:` statement here to make it explicit what you intend to code here.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        /// <summary>
+        /// Canvas on key up event
+        /// </summary>
+        /// <param name="e">the keyboard args of the event</param>
+        public void OnKeyUp(KeyboardEventArgs e)
+        {
+
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:150]: I propose you put a `//TODO:` statement here to make it explicit what you intend to code here.

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            Scene.AddPrimitive(xAxis, Color.Red);
+
+            Line yAxis = new Line(0, -size, 0, 0, size, 0);
+            Scene.AddPrimitive(yAxis, Color.Green);
+
+            Line zAxis = new Line(0, 0, -size, 0, 0, size);
+            Scene.AddPrimitive(zAxis, Color.Blue);
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:169]: add new line at the end of file

File: COMETwebapp/Primitives/Cone.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Cone.cs" company="RHEA System S.A.">
+//    Copyright (c) 2022 RHEA System S.A.
+//
+//    Author: Jaime Bernar
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cone.cs:5]: please add: , Sam Gerené, Alex Vorobiev, Alexander van Delft

File: COMETwebapp/Primitives/Cube.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Cube.cs" company="RHEA System S.A.">
+//    Copyright (c) 2022 RHEA System S.A.
+//
+//    Author: Jaime Bernar
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cube.cs:5]: please add: , Sam Gerené, Alex Vorobiev, Alexander van Delft

File: COMETwebapp/Primitives/CustomPrimitive.cs [hunk 9]
```diff
@@ -0,0 +1,57 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="CustomPrimitive.cs" company="RHEA System S.A.">
+//    Copyright (c) 2022 RHEA System S.A.
+//
+//    Author: Jaime Bernar
```
Comment (samatrhea) [inline COMETwebapp/Primitives/CustomPrimitive.cs:5]: please add: , Sam Gerené, Alex Vorobiev, Alexander van Delft

File: COMETwebapp/Primitives/CustomPrimitive.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        /// </summary>
+        /// <param name="path">the path to the file</param>
+        /// <param name="fileName">the file name with extension included</param>
+        public CustomPrimitive(string path, string fileName)
+        {
+            this.Path = path;
+            this.FileName = fileName;
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/CustomPrimitive.cs:57]: add new line at the end

File: COMETwebapp/Primitives/Cylinder.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Cylinder.cs" company="RHEA System S.A.">
+//    Copyright (c) 2022 RHEA System S.A.
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cylinder.cs:3]: please add: , Sam Gerené, Alex Vorobiev, Alexander van Delft

File: COMETwebapp/Primitives/Cylinder.cs [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        public Cylinder(double x, double y, double z, double radius, double height)
+        {
+            this.X = x;
+            this.Y = y;
+            this.Z = z;
+            this.Radius = radius;
+            this.Height = height;
+        }
+    }
+}
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Cylinder.cs:74]: add new line

File: COMETwebapp/Primitives/Line.cs [hunk 13]
```diff
@@ -0,0 +1,74 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Line.cs" company="RHEA System S.A.">
+//    Copyright (c) 2022 RHEA System S.A.
+//
+//    Author: Jaime Bernar
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Line.cs:5]: please add: , Sam Gerené, Alex Vorobiev, Alexander van Delft

File: COMETwebapp/Primitives/PositionablePrimitives.cs [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace COMETwebapp.Primitives
+{
+    public abstract class PositionablePrimitives : Primitive
```
Comment (samatrhea) [inline COMETwebapp/Primitives/PositionablePrimitives.cs:26]: please add XML docs

File: COMETwebapp/Primitives/PositionablePrimitives.cs [hunk 15]
```diff
@@ -0,0 +1,116 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="PositionablePrimitive.cs" company="RHEA System S.A.">
+//    Copyright (c) 2022 RHEA System S.A.
+//
+//    Author: Jaime Bernar
```
Comment (samatrhea) [inline COMETwebapp/Primitives/PositionablePrimitives.cs:5]: please add: , Sam Gerené, Alex Vorobiev, Alexander van Delft

File: COMETwebapp/Primitives/Primitive.cs [hunk 16]
```diff
@@ -0,0 +1,93 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Primitive.cs" company="RHEA System S.A.">
+//    Copyright (c) 2022 RHEA System S.A.
+//
+//    Author: Jaime Bernar
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Primitive.cs:5]: please add: , Sam Gerené, Alex Vorobiev, Alexander van Delft

File: COMETwebapp/Primitives/Sphere.cs [hunk 17]
```diff
@@ -0,0 +1,65 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Sphere.cs" company="RHEA System S.A.">
+//    Copyright (c) 2022 RHEA System S.A.
+//
+//    Author: Jaime Bernar
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Sphere.cs:5]: please add: , Sam Gerené, Alex Vorobiev, Alexander van Delft

File: COMETwebapp/Primitives/Torus.cs [hunk 18]
```diff
@@ -0,0 +1,74 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="Torus.cs" company="RHEA System S.A.">
+//    Copyright (c) 2022 RHEA System S.A.
+//
+//    Author: Jaime Bernar
```
Comment (samatrhea) [inline COMETwebapp/Primitives/Torus.cs:5]: please add: , Sam Gerené, Alex Vorobiev, Alexander van Delft

File: COMETwebapp/Scene.cs [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+    using System.Numerics;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Static class to acess the resources of a Scene 
+    /// </summary>
+    public static class Scene
+    {
+        /// <summary>
+        /// Collection of temporaty <see cref="Primitive"/> in the Scene.
```
Comment (samatrhea) [inline COMETwebapp/Scene.cs:40]: spellong -> temporaty: temporary

File: COMETwebapp/Scene.cs [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+    using COMETwebapp.Primitives;
+    using Newtonsoft.Json;
+    using System;
+    using System.Collections.Generic;
+    using System.Drawing;
+    using System.Numerics;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Static class to acess the resources of a Scene
```
Comment (samatrhea) [inline COMETwebapp/Scene.cs:35]: spelling `access`

File: COMETwebapp/Scene.cs [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace COMETwebapp
+{
+    using COMETwebapp.Primitives;
```
Comment (samatrhea) [inline COMETwebapp/Scene.cs:26]: please reorder using statements according to convention

File: COMETwebapp/wwwroot/Scripts/Helper.js [hunk 22]
```diff
@@ -0,0 +1,51 @@
+﻿function CreateMaterial(diffuse, specular, emissive, ambient, materialName, scene) {
```
Comment (samatrhea) [inline COMETwebapp/wwwroot/Scripts/Helper.js:1]: please add copyright header and document the methods

File: COMETwebapp/wwwroot/Scripts/babylonInterop.js [hunk 23]
```diff
@@ -0,0 +1,281 @@
+﻿"use strict";
```
Comment (samatrhea) [inline COMETwebapp/wwwroot/Scripts/babylonInterop.js:1]: please add copyright header and document the methods

File: COMETwebapp/Components/Viewer/BabylonCanvasBase.cs [hunk 24]
```diff
@@ -0,0 +1,169 @@
...
+        /// <summary>
+        /// Injected property to get acess to <see cref="ISessionAnchor"/>
+        /// </summary>
+        [Inject]
+        ISessionAnchor SessionAnchor { get; set; }
+
+        [JSInvokable]
+        public static string GetGUID() => Guid.NewGuid().ToString();
+
+        protected async override Task OnAfterRenderAsync(bool firstRender)
```
[the remark this answers is not in the document]
Comment (jaimeatrhea) [reply COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:36]: @samatrhea Can you explain me what do you mean?

[the remark this answers is not in the document]
Comment (alexatrhea) [reply COMETwebapp/Components/Viewer/BabylonCanvasBase.cs:36]: all code, like this method here should have documentation above it:

```
/// <summary>
/// ...
/// </summary>
```
etc

File: COMETwebapp/wwwroot/Scripts/babylonInterop.js [hunk 25]
```diff
@@ -0,0 +1,393 @@
...
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+//    Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+/// <summary>
```
Comment (samatrhea) [inline COMETwebapp/wwwroot/Scripts/babylonInterop.js:25]: please use javascript documentation style:

https://www.valentinog.com/blog/jsdoc/

File: tatic Dictionarystring, Primitive temporaryPrimitivesCollection = new Dictionarystring, Primitive(); [hunk 26]
```diff
@@ -0,0 +1,120 @@
+[33mc85bb2e[m[33m ([m[1;36mHEAD -> [m[1;32mfeat/GH159-Merge-BlazorViewer-COMET-WEB[m[33m, [m[1;31morigin/feat/GH159-Merge-BlazorViewer-COMET-WEB[m[33m)[m Added the Blazor Viewer; fixes #159
```
Comment (samatrhea) [inline tatic Dictionarystring, Primitive temporaryPrimitivesCollection = new Dictionarystring, Primitive();:1]: I don;t think this file needs to be in the code-base

**Review by samatrhea — CHANGES_REQUESTED** [review]

**Review by samatrhea — APPROVED** [review]

**Pull request merged**